payments-core/src/main/java/com/stripe/android/core/networking/DefaultStripeNetworkClient.kt

```diff
@@ -0,0 +1,102 @@
+package com.stripe.android.core.networking
+
+import androidx.annotation.VisibleForTesting
+import com.stripe.android.Logger
+import com.stripe.android.core.exception.APIConnectionException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.IOException
+import kotlin.coroutines.CoroutineContext
+
+internal class DefaultStripeNetworkClient @JvmOverloads constructor(
+    private val workContext: CoroutineContext = Dispatchers.IO,
+    private val connectionFactory: ConnectionFactory = ConnectionFactory.Default,
+    private val retryDelaySupplier: RetryDelaySupplier = RetryDelaySupplier(),
+    private val maxRetries: Int = DEFAULT_MAX_RETRIES,
+    private val logger: Logger = Logger.noop()
+) : StripeNetworkClient {
+    override suspend fun executeRequest(request: StripeRequest): StripeResponse<String> {
+        return executeInternal(maxRetries, request.retryResponseCodes) {
+            makeRequest(request)
+        }
+    }
+
+    override suspend fun executeRequestForFile(
+        request: StripeRequest,
+        outputFile: File
+    ): StripeResponse<File> {
+        return executeInternal(maxRetries, request.retryResponseCodes) {
+            makeRequestForFile(request, outputFile)
+        }
+    }
+
+    @VisibleForTesting
+    internal suspend fun <BodyType> executeInternal(
+        remainingRetries: Int,
+        retryResponseCodes: Iterable<Int>,
+        requester: () -> StripeResponse<BodyType>
+    ): StripeResponse<BodyType> = withContext(workContext) {
+
+        val stripeResponse = requester()
+
+        if (retryResponseCodes.contains(stripeResponse.code) && remainingRetries > 0) {
+            logger.info(
+                "Request was rate-limited with $remainingRetries remaining retries."
```

Comment:
If `retryResponseCodes` ever changes, this could be a misleading log line. Could we instead log something like

```
"Request failed with code ${stripeResponse.code}. Retrying up to $remainingRetries more time(s)."
```

Reply:
nice catch, updated, also added this in the test